File: BrainHatComponents/BrainflowInterfaces/Interfaces/CytonBoardSettings.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace BrainflowInterfaces
{
    public enum SrbSet
    {
        Unknown = -1,
        Disconnected,
        Connected,
    }


    public interface ICytonChannelSettings
    {
        int ChannelNumber { get; set; }
        bool PowerDown { get; set; }
        ChannelGain Gain { get; set; }
        AdsChannelInputType InputType { get; set; }
        bool Bias { get; set; }
        bool LlofP { get; set; }
        bool LlofN { get; set; }
        bool Srb2 { get; set; }

    }
    //
    public class CytonChannelSettingsImplementation : ICytonChannelSettings
    {
        public CytonChannelSettingsImplementation()
        {
            ChannelNumber = 0;
            PowerDown = false;
            Gain = ChannelGain.x24;
            InputType = AdsChannelInputType.Normal;
            Bias = true;
            Srb2 = true;
        }

        public int ChannelNumber { get; set; }
        public bool PowerDown { get; set; }
        public ChannelGain Gain { get; set; }
        public AdsChannelInputType InputType { get; set; }
        public bool Bias { get; set; }
        public bool LlofP { get; set; }
        public bool LlofN { get; set; }
        public bool Srb2 { get; set; }

        public override string ToString()
        {
            return this.ChannelSettingsToString();
        }
    }

    public interface ICytonBoardSettings
    {
        bool Srb1Set { get; }
        ICytonChannelSettings[] Channels { get; }
    }
    //
    public class CytonBoardSettingsImplementation : ICytonBoardSettings
    {
        public CytonBoardSettingsImplementation()
        {
            _Channels = new List<ICytonChannelSettings>();
        }

        public bool Srb1Set { get; set; }
        List<ICytonChannelSettings> _Channels;
        public ICytonChannelSettings[] Channels => _Channels.ToArray();
        public void AddChannel(ICytonChannelSettings channel)
        {
            _Channels.Add(channel);
        }

    }

    public interface ICytonBoards
    {
        ICytonBoardSettings[] Boards { get; }
        bool IsValid { get; }
    }
    //
    public class CytonBoardsImplementation : ICytonBoards
    {
        public CytonBoardsImplementation()
        {
            _Boards = new List<CytonBoardSettingsImplementation>();
        }

        List<CytonBoardSettingsImplementation> _Boards;
        public ICytonBoardSettings[] Boards => _Boards.ToArray();

        public bool IsValid => _Boards.Count > 0;

        public void Invalidate()
        {
            _Boards = new List<CytonBoardSettingsImplementation>();
        }
        public void LoadFromRegistersString(string registerReport)
        {
            registerReport = registerReport.Replace('\r', ' ');
            registerReport = registerReport.Replace(" ", String.Empty);

            _Boards = new List<CytonBoardSettingsImplementation>();

            int boardChannelOffset = 0;

            try
            {
                var lines = registerReport.Split('\n');
                foreach (var nextLine in lines)
                {
                    if (nextLine.Contains("ADSRegisters"))
                    {
                        if (_Boards.Count > 0)
                        {
                            if (_Boards[_Boards.Count - 1].Channels.Count() != 8)
                            {
                                _Boards = new List<CytonBoardSettingsImplementation>();
                                throw new Exception("Board has less than 8 channels");
                            }
                            boardChannelOffset += _Boards.Last().Channels.Count();
                        }

                        _Boards.Add(new CytonBoardSettingsImplementation());
                    }
                    else if (nextLine.Length > "CHx".Length && nextLine.Substring(0, "CH".Length) == "CH")
                    {
                        var columns = nextLine.Split(',');
                        if (columns.Count() == 11)
                        {
                            var newChannel = new CytonChannelSettingsImplementation();

                            newChannel.ChannelNumber = int.Parse(nextLine.Substring("CH".Length, 1)) + boardChannelOffset;

                            if (_Boards.Last().Channels.Length > 0)
                            {
                                if (newChannel.ChannelNumber - _Boards.Last().Channels.Last().ChannelNumber != 1)
                                {
                                    _Boards = new List<CytonBoardSettingsImplementation>();
                                    throw new Exception("Board channels are not sequential");
                                }
                            }

                            newChannel.PowerDown = columns[7.IndexOfBit()] == "1" ? true : false;
                            newChannel.Gain = columns.GetChannelGain();
                            newChannel.InputType = columns.GetChannelInputType();
                            newChannel.Srb2 = columns[3.IndexOfBit()] == "1" ? true : false;

                            _Boards.Last().AddChannel(newChannel);

                        }
                    }
                    else if (nextLine.Length > "BIAS_SENSP".Length && nextLine.Substring(0, "BIAS_SENSP".Length) == "BIAS_SENSP")
                    {
                        ValidateBoardChannels();

                        var columns = nextLine.Split(',');
                        if (columns.Length == 11)
                        {
                            int bit = 0;
                            foreach (var nextChannel in _Boards.Last().Channels)
                            {
                                nextChannel.Bias = columns[bit.IndexOfBit()] == "1" ? true : false;
                                bit++;
                            }
                        }
                    }
                    else if (nextLine.Length > "LOFF_SENSP".Length && nextLine.Substring(0, "LOFF_SENSP".Length) == "LOFF_SENSP")
                    {
                        ValidateBoardChannels();

                        var columns = nextLine.Split(',');
                        if (columns.Length == 11)
                        {
                            int bit = 0;
                            foreach (var nextChannel in _Boards.Last().Channels)
                            {
                                nextChannel.LlofP = columns[bit.IndexOfBit()] == "1" ? true : false;
                                bit++;
                            }
                        }
                    }
                    else if (nextLine.Length > "LOFF_SENSN".Length && nextLine.Substring(0, "LOFF_SENSN".Length) == "LOFF_SENSN")
                    {
                        ValidateBoardChannels();

                        var columns = nextLine.Split(',');
                        if (columns.Length == 11)
                        {
                            int bit = 0;
                            foreach (var nextChannel in _Boards.Last().Channels)
                            {
                                nextChannel.LlofN = columns[bit.IndexOfBit()] == "1" ? true : false;
                                bit++;
                            }
                        }
                    }
                    else if (nextLine.Length > "MISC1".Length && nextLine.Substring(0, "MISC1".Length) == "MISC1")
                    {
                        ValidateBoardChannels();

                        var columns = nextLine.Split(',');
                        if (columns.Length == 11)
                        {
                            _Boards.Last().Srb1Set = columns[5.IndexOfBit()] == "1" ? true : false;
                        }
                    }
                }


            }
            catch (Exception e)
            {
                throw e;
            }
        }

        void ValidateBoardChannels()
        {
            if (_Boards.Last().Channels.Length != 8)
            {
                _Boards = new List<CytonBoardSettingsImplementation>();
                throw new Exception("Fewer than 8 channels in the board");
            }
        }




    }

    public static class CytonBoardExtensionMethods
    {
        public static int IndexOfBit(this int value)
        {
            return 10 - value;
        }

        public static ChannelGain GetChannelGain(this string[] value)
        {
            if (value.Count() == 11)
            {
                int channelGain = 0;
                channelGain += value[4.IndexOfBit()] == "1" ? 1 : 0;
                channelGain += value[5.IndexOfBit()] == "1" ? 2 : 0;
                channelGain += value[6.IndexOfBit()] == "1" ? 4 : 0;

                return (ChannelGain)channelGain;
            }

            return ChannelGain.x1;
        }

        public static AdsChannelInputType GetChannelInputType(this string[] value)
        {
            if (value.Count() == 11)
            {
                int channelGain = 0;
                channelGain += value[0.IndexOfBit()] == "1" ? 1 : 0;
                channelGain += value[1.IndexOfBit()] == "1" ? 2 : 0;
                channelGain += value[2.IndexOfBit()] == "1" ? 4 : 0;

                return (AdsChannelInputType)channelGain;
            }

            return AdsChannelInputType.Normal;
        }
    }
}

File: brainHatServer/brainHat/CytonBoardSettings.cpp
#include "brainHat.h"
#include "CytonBoardSettings.h"
#include "StringExtensions.h"

using namespace std;


// Cyton Board Settings
//
#pragma  region CytonBoard

CytonBoardSettings::CytonBoardSettings()
{
	Srb1Set = false;
}
	

//  Destructor
//
CytonBoardSettings::~CytonBoardSettings()
{
	for (auto it = Channels.begin(); it != Channels.end(); ++it)
	{
		delete *it;
	}
	Channels.clear();
}
	
	
//  Add a channel to the channels collection	
//
void CytonBoardSettings::AddChannel(CytonChannelSettings* channelSettings)
{
	Channels.push_back(channelSettings);
}
#pragma endregion



//  CytonBoards
//
#pragma region CytonBoards

CytonBoards::CytonBoards()
{
}

// Destructor
//
CytonBoards::~CytonBoards()
{
	ClearBoards();
}
	

//  Clear Boards
//
void CytonBoards::ClearBoards()
{
	for (auto it = Boards.begin(); it != Boards.end(); ++it)
	{
		delete *it;
	}
	Boards.clear();
}


//  Read from register string
//
bool CytonBoards::ReadFromRegisterString(string registerReport)
{
	registerReport.erase(remove(registerReport.begin(), registerReport.end(), '\r'), registerReport.end());
	registerReport.erase(remove(registerReport.begin(), registerReport.end(), ' '), registerReport.end());
	
	int boardChannelOffset = 0;
	
	string CHx = "CHx";
	string CH = "CH";
	string BIAS_SENSP = "BIAS_SENSP";
	string LOFF_SENSP = "LOFF_SENSP";
	string LOFF_SENSN = "LOFF_SENSN";
	string MISC1 = "MISC1";
		
	vector<string> lines;
	Tokenize(registerReport, lines, "\n");
		
	for (auto it = lines.begin(); it != lines.end(); ++it)
	{
		auto nextLine = *it;
		if (nextLine.find("ADSRegisters") != string::npos)
		{
			if (Boards.size() > 0)
			{
				if (Boards[Boards.size() - 1]->Channels.size() != 8)
				{
					Logging.AddLog("CytonBoardSettings", "ReadFromRegisterString", "Board does not have 8 channels.", LogLevelError);
					ClearBoards();
					return false;
				}
				boardChannelOffset += Boards.back()->Channels.size();
			}

			Boards.push_back(new CytonBoardSettings());
		}
		else if (nextLine.length() > CHx.length() && nextLine.substr(0, CH.length()) == CH)
		{
			vector<string> columns;
			Tokenize(nextLine, columns, ",");
			if (columns.size() == 11)
			{
				auto newChannel = new CytonChannelSettings();
				
				newChannel->ChannelNumber = ParseInt(nextLine.substr(CH.length(), 1)) + boardChannelOffset;
                           
				if (Boards.back()->Channels.size() > 0)
				{
					if (newChannel->ChannelNumber - Boards.back()->Channels.back()->ChannelNumber != 1)
					{
						Logging.AddLog("CytonBoardSettings", "ReadFromRegisterString", "Board channels are not sequential.", LogLevelError);
						ClearBoards();
						delete newChannel;
						return false;
					}
				}
                            
				newChannel->PowerDown = columns[IndexOfBit(7)] == "1" ? true : false;
				newChannel->Gain = GetChannelGain(columns);
				newChannel->InputType = GetChannelInputType(columns);
				newChannel->Srb2 = columns[IndexOfBit(3)] == "1" ? true : false;
				Boards.back()->AddChannel(newChannel);
			}
		}
		else if (nextLine.length() > BIAS_SENSP.length() && nextLine.substr(0, BIAS_SENSP.length()).compare(BIAS_SENSP) == 0)
		{
			if (!ValidateBoardChannels())
			{
				Logging.AddLog("CytonBoardSettings", "ReadFromRegisterString", "Board does not have 8 channels.", LogLevelError);
				ClearBoards();
				return false;
			}

			vector<string> columns;
			Tokenize(nextLine, columns, ",");
			if (columns.size() == 11)
			{
				int bit = 0;
				for (auto nextChannel = Boards.back()->Channels.begin(); nextChannel != Boards.back()->Channels.end(); ++nextChannel) 
				{
					(*nextChannel)->Bias = columns[IndexOfBit(bit)] == "1" ? true : false;
					bit++;
				}
			}
		}
		else if (nextLine.length() > LOFF_SENSP.length() && nextLine.substr(0, LOFF_SENSP.length()).compare(LOFF_SENSP) == 0)
		{
			if (!ValidateBoardChannels())
			{
				Logging.AddLog("CytonBoardSettings", "ReadFromRegisterString", "Board does not have 8 channels.", LogLevelError);
				ClearBoards();
				return false;
			}

			vector<string> columns;
			Tokenize(nextLine, columns, ",");
			if (columns.size() == 11)
			{
				int bit = 0;
				for (auto nextChannel = Boards.back()->Channels.begin(); nextChannel != Boards.back()->Channels.end(); ++nextChannel) 
				{
					(*nextChannel)->LlofP = columns[IndexOfBit(bit)] == "1" ? true : false;
					bit++;
				}
			}
		}
		else if (nextLine.length() > LOFF_SENSN.length() && nextLine.substr(0, LOFF_SENSN.length()).compare(LOFF_SENSN) == 0)
		{
			if (!ValidateBoardChannels())
			{
				Logging.AddLog("CytonBoardSettings", "ReadFromRegisterString", "Board does not have 8 channels.", LogLevelError);
				ClearBoards();
				return false;
			}

			vector<string> columns;
			Tokenize(nextLine, columns, ",");
			if (columns.size() == 11)
			{
				int bit = 0;
				for (auto nextChannel = Boards.back()->Channels.begin(); nextChannel != Boards.back()->Channels.end(); ++nextChannel) 
				{
					(*nextChannel)->LlofN = columns[IndexOfBit(bit)] == "1" ? true : false;
					bit++;
				}
			}
		}
		else if (nextLine.length() > MISC1.length() && nextLine.substr(0, MISC1.length()).compare(MISC1) == 0)
		{
			if (!ValidateBoardChannels())
			{
				Logging.AddLog("CytonBoardSettings", "ReadFromRegisterString", "Board does not have 8 channels.", LogLevelError);
				ClearBoards();
				return false;
			}

			vector<string> columns;
			Tokenize(nextLine, columns, ",");
			if (columns.size() == 11)
			{
				Boards.back()->Srb1Set = columns[IndexOfBit(5)] == "1" ? true : false;
			}
		}
	}
	
	return true;

   
	
}


bool CytonBoards::ValidateBoardChannels()
{
	return Boards.back()->Channels.size() == 8;
}
	
#pragma endregion
